### PR TITLE
Introduce export and import for global shortcuts

### PR DESCRIPTION
**Pull request by Kissaki**
This functionality ignores server specific shortcuts (shortcuts with
server specific targets).
These are stored to the database through a different mechanism.

This feature was requested on the forums in [Mumble shortcuts export/import](http://forums.mumble.info/viewtopic.php?f=4&t=3367)

2018-01-06 15_22_10-mumble configuration

### CODE REVIEW

File: src/mumble/Settings.cpp [hunk 1]
```diff
@@ -825,6 +811,26 @@ void Settings::load(QSettings* settings_ptr) {
...
 }
 
+QList<Shortcut> Settings::loadShortcuts(QSettings *settings_ptr) {
+	QList<Shortcut> qlShortcuts;
+	int nshorts = settings_ptr->beginReadArray(QLatin1String("shortcuts"));
+	for (int i=0; i<nshorts; i++) {
+		settings_ptr->setArrayIndex(i);
+		Shortcut s;
+
+		s.iIndex = -2;
```
Comment (hacst) [inline src/mumble/Settings.cpp]: Why -2 ?

File: src/mumble/GlobalShortcut.ui [hunk 2]
```diff
@@ -204,6 +204,26 @@
           </property>
          </spacer>
         </item>
+        <item>
+         <widget class="QPushButton" name="qpbImport">
+          <property name="toolTip">
+           <string>Import shortcuts from a file, replacing the current shortcuts.</string>
```
Comment (hacst) [inline src/mumble/GlobalShortcut.ui]: Simplified English suggestion:
"Import shortcuts from a file. Overwrites the current shortcuts."

File: src/mumble/GlobalShortcut.cpp [hunk 3]
```diff
@@ -625,6 +625,27 @@ void GlobalShortcutConfig::on_qpbRemove_clicked(bool) {
...
+}
+
+void GlobalShortcutConfig::on_qpbExport_clicked() {
+	commit();
+	QString filePath = QFileDialog::getSaveFileName(this, tr("Save shortcuts"), QLatin1String("mumble-shortcuts.ini"), QLatin1String("*.ini"));
+	if (filePath.isEmpty()) {
+		return;
+	}
+	QSettings s(filePath, QSettings::IniFormat);
+	g.s.saveShortcuts(qlShortcuts, &s);
```
Comment (hacst) [inline src/mumble/GlobalShortcut.cpp]: Should sync and then check status (see http://doc.qt.io/qt-5/qsettings.html#status) to make sure we were actually able to write.

File: src/mumble/GlobalShortcut.cpp [hunk 4]
```diff
@@ -625,6 +625,27 @@ void GlobalShortcutConfig::on_qpbRemove_clicked(bool) {
...
 }
 
+void GlobalShortcutConfig::on_qpbImport_clicked() {
+	commit();
+	QString filePath = QFileDialog::getOpenFileName(this, tr("Import shortcuts"), QString(), QLatin1String("*.ini"));
+	if (filePath.isEmpty()) {
+		return;
+	}
+	QSettings s(filePath, QSettings::IniFormat);
+	qlShortcuts = g.s.loadShortcuts(&s);
```
Comment (hacst) [inline src/mumble/GlobalShortcut.cpp]: Not ideal. If we are passed a non-existent / invalid file we will nuke the users shortcuts without being able to say what happened. Should at least try whether we can use the status() function to verify the ini didn't contain any obvious errors. To make it actually robust we probably have to write/load an additional key in the export/import to indicate this is a mumble shortcut file. That way the user cannot accidentally select some random ini and we can notify properly if we end up with an invalid file. If we make it a versiony thing it might come in handy in the future.

File: src/mumble/Settings.cpp [hunk 5]
```diff
@@ -825,6 +811,26 @@ void Settings::load(QSettings* settings_ptr) {
 	settings_ptr->endGroup();
 }
 
+QList<Shortcut> Settings::loadShortcuts(QSettings *settings_ptr) {
+	QList<Shortcut> qlShortcuts;
+	int nshorts = settings_ptr->beginReadArray(QLatin1String("shortcuts"));
```
Comment (hacst) [inline src/mumble/Settings.cpp]: CG: const

File: src/mumble/Settings.cpp [hunk 6]
```diff
@@ -825,6 +811,26 @@ void Settings::load(QSettings* settings_ptr) {
...
+	for (int i=0; i<nshorts; i++) {
```
Comment (hacst) [inline src/mumble/Settings.cpp]: CG: spacing

File: src/mumble/GlobalShortcut.cpp [hunk 7]
```diff
@@ -625,6 +625,27 @@ void GlobalShortcutConfig::on_qpbRemove_clicked(bool) {
...
+	reload();
+}
+
+void GlobalShortcutConfig::on_qpbExport_clicked() {
+	commit();
+	QString filePath = QFileDialog::getSaveFileName(this, tr("Save shortcuts"), QLatin1String("mumble-shortcuts.ini"), QLatin1String("*.ini"));
```
Comment (hacst) [inline src/mumble/GlobalShortcut.cpp]: CG: const

File: src/mumble/GlobalShortcut.cpp [hunk 8]
```diff
@@ -625,6 +625,27 @@ void GlobalShortcutConfig::on_qpbRemove_clicked(bool) {
 	qlShortcuts.removeAt(idx);
 }
 
+void GlobalShortcutConfig::on_qpbImport_clicked() {
+	commit();
+	QString filePath = QFileDialog::getOpenFileName(this, tr("Import shortcuts"), QString(), QLatin1String("*.ini"));
```
Comment (hacst) [inline src/mumble/GlobalShortcut.cpp]: CG: const

Comment (hacst) [inline src/mumble/GlobalShortcut.cpp]: Imho should warn before overwriting all existing shortcuts. We can't expect the user to have read the tooltip and we don't have an undo button. Doesn't necessarily have to be a separate dialog box. Maybe put it in the selector title?

File: src/mumble/GlobalShortcut.cpp (hunk 4 above)
Comment (hacst) [inline src/mumble/GlobalShortcut.cpp]: Won't doing it this way also nuke the server specific shortcuts?

**Comment by Kissaki** [comment]
Buttons are in the 'action bar' on the right.
2018-01-06 15_22_10-mumble configuration

File: src/mumble/Settings.cpp (hunk 1 above)
Comment (Kissaki) [reply src/mumble/Settings.cpp]: That’s just moved code. It was `-2` before.

Can’t quickly see what it even is…

File: src/mumble/Settings.h [hunk 9]
```diff
@@ -411,7 +411,9 @@ struct Settings {
 	Settings();
 	void load();
 	void load(QSettings*);
+	QList<Shortcut> loadShortcuts(QSettings *s);
```
[the remark this answers is not in the document]
Comment (Kissaki) [reply src/mumble/Settings.h]: Not sure about noting external use. By that reasoning we would have to document every public method on external use.

File: src/mumble/GlobalShortcut.cpp (hunk 4 above)
Comment (Kissaki) [reply src/mumble/GlobalShortcut.cpp]: The `SAVELOAD` makro uses `settings_ptr`…

File: src/mumble/GlobalShortcut.cpp (hunk 8 above)
Comment (Kissaki) [reply src/mumble/GlobalShortcut.cpp]: I think the title would not be enough. It would have to be an adequately named action button, which is not feasible with standard forms.
Warning in a dialog box seems better.

File: src/mumble/GlobalShortcut.cpp (hunk 4 above)
Comment (Kissaki) [reply src/mumble/GlobalShortcut.cpp]: I don’t remember…
Export says it ignores server specific shortcuts, and that just iterates only `qlShortcuts`.

Comment (Kissaki) [reply src/mumble/GlobalShortcut.cpp]: Yes it does. Shortcuts are added externally into `qlShortcuts`.

**Comment by Kissaki** [comment]
@hacst PTAL

File: src/mumble/GlobalShortcut.cpp [hunk 10]
```diff
@@ -625,6 +625,39 @@ void GlobalShortcutConfig::on_qpbRemove_clicked(bool) {
...
 
+void GlobalShortcutConfig::on_qpbImport_clicked() {
+	commit();
+	const QString filePath = QFileDialog::getOpenFileName(this, tr("Import shortcuts"), QString(), QLatin1String("*.ini"));
+	if (filePath.isEmpty()) {
+		return;
+	}
+	int userAction = QMessageBox::warning(this
+	                                      , tr("Shortcut Replacement")
+	                                      , tr("Importing the shortcuts will drop and replace all current shortcuts.\nDo you want to continue?")
```
Comment (hacst) [inline src/mumble/GlobalShortcut.cpp]: Maybe clarify "all current shortcuts including server specific ones". Could also suggest to export the current ones if you don't want to loose them.

File: src/mumble/GlobalShortcut.cpp [hunk 11]
```diff
@@ -625,6 +625,39 @@ void GlobalShortcutConfig::on_qpbRemove_clicked(bool) {
...
+	                                      , QMessageBox::Ok | QMessageBox::Cancel
+	                                      , QMessageBox::Ok);
+	if (userAction != QMessageBox::Ok) {
+		return;
+	}
+	QSettings s(filePath, QSettings::IniFormat);
+	qlShortcuts = Settings::loadShortcuts(&s);
```
Comment (hacst) [inline src/mumble/GlobalShortcut.cpp]: Doesn't this still allow you to accidentally select a random file that isn't exported shortcuts and have your shortcuts be nuked without any feedback? See previous review comment.

**Comment by Kissaki** [comment]
@hacst PTAL

**Comment by mkrautz** [comment]
@Kissaki 

One thing to be aware of is that we don't really have cross-platform way of representing shortcuts. That is to say, the internal representation is dependent on the global shorcut backend we're using. That also means that it might not work the same way on different Linux boxes!

For example, I don't think evdev on Linux is compatible with XInput2. If we add Wayland support, that would use different values yet again. Same with game controllers.

That is problemnatic for this feature, but I dont' think it should block your work. The import/export can work without us "stabilizing" the internal interface.

**Comment by mkrautz** [comment]
A follow-up:

In practice, the situation isn't as dire as I presented it.

On Windows, we use multiple backends. Winhooks and DirectInput. We map the WinHooks data to DirectInput, such that the two are cross-compatible. We also support other backends, such as GKeys and XboxInput. Those are serialized as specific to those backends, since they're quite specialized.

On Linux, we support XInput2 and evdev. But both use the same button mapping: https://github.com/mumble-voip/mumble/blob/f6ba3a5/src/mumble/GlobalShortcut_unix.cpp#L331, which suggests that the keycodes match between XInput and evdev. So not much of a problem there.

On macOS, we only support Quartz Event Tap. No cross-compatibility.

TL;DR: For the same OS, exported shortcuts should be compatible.
However, importing a Windows shortcut file will not work on macOS, nor Linux. And vice versa.

Maybe it'd be sensible to include an OS watermark, like you do for the version?

File: src/mumble/Settings.cpp [hunk 12]
```diff
@@ -778,21 +780,7 @@ void Settings::load(QSettings* settings_ptr) {
...
-
-		SAVELOAD(s.iIndex, "index");
-		SAVELOAD(s.qlButtons, "keys");
-		SAVELOAD(s.bSuppress, "suppress");
-		s.qvData = settings_ptr->value(QLatin1String("data"));
-		if (s.iIndex >= -1)
-			qlShortcuts << s;
-	}
-	settings_ptr->endArray();
+	qlShortcuts = loadShortcuts(settings_ptr);
```
Comment (hacst) [inline src/mumble/Settings.cpp]: Won't this fail to load the shortcuts now that the loadShortcuts function checks for fileVersion? The registry won't have that on first launch after update or am I missing something?

If that is the case my suggestion would be to have the loadShortcuts/saveShortcuts function be the normal load/save functionality and introduce a thin wrapper loadShortcutsFile/saveShortcutsFile that does the version safe/restore + file handling. That way we don't mix the file save/load specific logic with our normal registry operation while still preventing code-duplication.

Comment (Kissaki) [reply src/mumble/Settings.cpp]: Yeah, I fucked that up…

**Comment by Kissaki** [comment]
Yeah, I guess I’ll have to also add a backend/data type qualifier.
Even if the OS would currently be sufficient as a distinguisher, I’d rather use the backend type and make cross-compatibilities explicitly allowed.

**Comment by Krzmbrzl** [comment]
Did you get to work on this some more? From reading the comments, it seems as if this PR is almost done :)

**Comment by Kissaki** [comment]
I do not have a working build environment so I have not worked on any code nor this PR.

**Comment by Krzmbrzl** [comment]
Closing this as abandoned
